src/home-staging/replace-furniture.js

```diff
@@ -0,0 +1,139 @@
+import getFurnitureInfo  from '../furniture/get-info.js'
+import callService  from '../utils/services/call.js'
+import normalizeFurnitureInfo from '../furniture/common/normalize-furniture-info.js'
+
+var userQuery, searchCount, margin, furnitureInfo, position, rotation
+
+var config = {
+  'default_margin': 0.1,
+  'default_search': '-generic isPublished:true',
+  'tag_black_list': [
+    'simplygon',
+    'hasChangeableMaterials',
+    'autofurnish',
+    'wallAttached',
+    'daybed',
+    'oval',
+    'rectangular',
+    '2 seater',
+    '3 seater',
+    '4 seater'
+  ],
+}
+
+export default function replaceFurniture (id, options) {
+  // API
+  options = options || {}
+  userQuery = options.query || null
+  position = options.position || {x: 0, y: 0, z: 0}
+  rotation = options.rotation || {x: 0, y: 0, z: 0}
+  // config publishable api key
+  // reject when no publishable or not white listed domain
+  // we need to call furniture info first in order to obtain data3d URL
+  return getFurnitureInfo(id)
+    .then(function(info){
+      furnitureInfo = info
+      margin = config['default_margin']
+      searchCount = 0
+      var searchQuery = getQuery(furnitureInfo)
+      return search(searchQuery)
+    })
+    .then(function(result) {
+      return verifyResult(result, id)
+    })
+    .catch(function(error) {
+      console.error(error.message)
+    })
+}
+
+function verifyResult(result, id) {
+  if (searchCount > 10 ) {
+    return Promise.reject(new Error('No furniture was found'))
+  }
+  var rawResult = result.filter(function(el){
+    return el.productResourceId !== id
+  });
+
+  if (rawResult.length < 2) {
+    margin += 0.10
+    searchCount += 1
+    var searchQuery = getQuery(furnitureInfo);
+    return search(searchQuery).then(function(result) {
+      return verifyResult(result, id)
+    })
+  } else {
+    var cleanResult = rawResult.map(normalizeFurnitureInfo).map(function(res) {
+      return {
+        furniture: res,
+        position: computeNewPosition(furnitureInfo, res)
+      }
+    })
+    return Promise.resolve(cleanResult)
+  }
+}
+
+function search(searchQuery) {
+  return callService('Product.search', {searchQuery: searchQuery, limit: 200})
+}
+
+function getQuery(info) {
+  var query = config['default_search']
+  var tags = info.tags
+
+  tags.forEach(function(tag) {
+    // filter out black listed tags
+    if (config['tag_black_list'].indexOf(tag) < 0 ) {
```

Comment:
This can be written in a more readable form:

```javascript
tags.filter(function(tag) {
  // removes blacklisted tags as well as 1P, 2P, ...
  return !config['tag_black_list'].includes(tag) && !/^\d+P$/.test(tag)
}
```

`someArr.filter(function(elem) {})` will return an array that only contains elements where the function returned true :-)